### PR TITLE
Fetch deals

### PR DESCRIPTION
**Pull request by maitnngo2002**

### CODE REVIEW

File: WiseBuy/Managers/APIManager.m [hunk 1]
```diff
@@ -24,16 +45,140 @@ + (void)fetchDealsFromEbayAPI {
...
+    
+    NSString *const path = [[NSBundle mainBundle] pathForResource: @"Keys" ofType: @"plist"];
+    NSDictionary *const dict = [NSDictionary dictionaryWithContentsOfFile: path];
+    NSString *userKey = [dict objectForKey: @"searchUPC_userKey"];
+    
+    if ([[NSUserDefaults standardUserDefaults] stringForKey:@"upcDatabase_userKey"]) {
+        userKey = [[NSUserDefaults standardUserDefaults] stringForKey:@"upcDatabase_userKey"];
+    }
+    NSDictionary *headers = @{
+      @"user_key": @"38ece1f67747388f5034080aeebc2dc9"
```
Comment (jl345) [inline WiseBuy/Managers/APIManager.m:81]: Should this be here?

**Review by jl345 — COMMENTED** [review]
Thanks for creating a PR! It makes it easier for me to review logical portions as you write them. Some high level thoughts:

- The APIManager makes sense. It could be more scalable/maintainable by changing how it's structured, as I mentioned over chat
- The DatabaseManager was harder to follow. I'll take another look at it later and send the comments that I have for the rest of the code first
- This PR is large. I understand that since you're writing a lot of new code, it's tougher to split it up into small PRs (we can talk about what "small" means in more detail). For this one, some places where it could have been split up:
  - Adding one remote API at a time rather than all three
  - Making a PR with only the model changes. These are straightforward so they're super easy to review
  - Database manager changes (maybe even more than one PR, as there are multiple new methods. It doesn't have to be one PR per method though)
  - AlertManager changes
- If you are trying to build complete features in each PR, you may not be able to split up the PRs as I just described without missing functionality. It's okay for things to be incomplete--you can use TODO comments to indicate when something will be filled in later, and (for example) you can have placeholder methods which return nil or do nothing. When working on production code at Meta, new features are often handled in two different ways: either you make a stack of changes and you don't merge them until the feature is complete (this would be equivalent to making several PRs against a large Git feature branch, and only merging the branch at the end), or you hide new/incomplete features using feature flags so that the public doesn't see them. This is too much overhead for your project, so I think it's okay to treat the entire project as not publicly visible yet
- It looks like there's some stuff committed that shouldn't be. Be sure to update your .gitignore

As meta-feedback on the PR itself, the three things you would typically need for a PR at Meta are a title, a description, and a test plan. The test plan would have instructions for how you tested it and how someone reviewing the code could reproduce, or screenshots of a feature working. For your project, you don't have to include test plans for every PR (it introduces a lot of overhead), but you should include good titles and descriptions. In this case, "fetch deals" is a bit broad to cover everything this PR does, as seen by my bullet points above. In fact, the fact that it is hard to describe everything this PR does in a short title is a clue that it should be split into multiple PRs. There are a lot of resources for how to write good titles and descriptions online (one example: https://gist.github.com/robertpainsi/b632364184e70900af4ab688decf6f53, which was the first result I saw on Google). Don't worry about the style rules like the 72 character limit, but the rest of the advice is good.

File: WiseBuy/Managers/APIManager.m [hunk 2]
```diff
@@ -24,16 +43,125 @@ + (void)fetchDealsFromEbayAPI {
...
+            NSLog(@"%@",responseDictionary);
+            dispatch_semaphore_signal(sema);
+            
+      }
+    }];
+    [dataTask resume];
+    dispatch_semaphore_wait(sema, DISPATCH_TIME_FOREVER);
+}
+
++ (void)fetchDealsFromUPCDatabase{
```
Comment (jl345) [inline WiseBuy/Managers/APIManager.m]: (I thought I wrote a comment about this, but I don't see it now...re-writing it)

Having one method per API is fine to start, but as a suggestion for a future improvement, think about how to split up this code to make it more scalable when multiple APIs are added. Putting the APIs all into this one file would make this file gigantic, and is hard to maintain with multiple developers working at the same time.

File: WiseBuy/Keys.plist [hunk 3]
```diff
@@ -2,6 +2,14 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+	<key>access_Token</key>
```
[the remark this answers is not in the document]
Comment (maitnngo2002) [reply WiseBuy/Keys.plist:5]: have this fixed in the newer commits!

File: WiseBuy/Managers/APIManager.m [hunk 4]
```diff
@@ -8,12 +8,33 @@
...
+
+    NSString *const path = [[NSBundle mainBundle] pathForResource: @"Keys" ofType: @"plist"];
+    NSDictionary *const dict = [NSDictionary dictionaryWithContentsOfFile: path];
+    NSString *securityName = [dict objectForKey: @"appSecurityName"];
+    
+    if ([[NSUserDefaults standardUserDefaults] stringForKey:@"appSecurityName"]) {
+        securityName = [[NSUserDefaults standardUserDefaults] stringForKey:@"appSecurityName"];
+    }
+    
+    NSString *appSecurityName = securityName;
```
[the remark this answers is not in the document]
Comment (maitnngo2002) [reply WiseBuy/Managers/APIManager.m:29]: this reassigning is just to make it closely match the actual string itself

File: WiseBuy/Managers/APIManager.m [hunk 5]
```diff
@@ -8,12 +8,33 @@
...
+    NSString *firstHalfBaseURL = @"https://svcs.ebay.com/services/search/FindingService/v1?OPERATION-NAME=findItemsByProduct&SERVICE-VERSION=1.0.0&SECURITY-APPNAME=";
```
[the remark this answers is not in the document]
Comment (maitnngo2002) [reply WiseBuy/Managers/APIManager.m:30]: already have this fixed in the newer commits when refactoring code

File: WiseBuy/Managers/APIManager.m (hunk 1 above)
[the remark this answers is not in the document]
Comment (maitnngo2002) [reply WiseBuy/Managers/APIManager.m:81]: it have been fixed.

File: WiseBuy/Managers/DatabaseManager.m [hunk 6]
```diff
@@ -38,11 +42,176 @@ +(void)registerUser:(User *)user withCompletion:(void(^)(BOOL success, NSError *
     }];
 }
 
-+(PFFileObject *) getPFFileFromImage: (NSData *)imageData {
++ (void)fetchItem:(NSString *)barcode viewController:(UIViewController *)vc withCompletion:(void(^)(NSArray *deals,NSError *error))completion {
+    PFQuery *itemQuery = [Item query];
+    [itemQuery whereKey:@"barcode" equalTo:@"085239058152"];
```
[the remark this answers is not in the document]
Comment (maitnngo2002) [reply WiseBuy/Managers/DatabaseManager.m:47]: already practicing adding` // TODO: ` for remain-to-be-fixed bugs

**Pull request merged**